components/fxa-client/ios/FxAClient/FxAccountDeviceConstellation.swift

```diff
@@ -0,0 +1,192 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+public protocol DeviceConstellationProtocol {
+    // Get local + remote devices.
+    func state() -> ConstellationState?
+    // Refresh the list of remote devices.
+    // Dispatchs a notification to NSNotificationCenter, using the name `.constellationStateUpdate`.
+    // The updated `ConstellationState` can be found in `userInfo` under the key `newState`.
+    func refreshState()
+
+    func setLocalDeviceName(name: String)
+
+    // Poll for events we might have missed (e.g. no push notification)
+    func pollForEvents(completionHandler: @escaping (Result<[DeviceEvent], Error>) -> Void)
+    // Send an event to another device such as Send Tab.
+    func sendEventToDevice(targetDeviceId: String, e: DeviceEventOutgoing)
+
+    // Register our push subscription with the FxA server.
+    func setDevicePushSubscription(sub: DevicePushSubscription)
+    // Used by Push when receiving a message.
+    func processRawIncomingDeviceEvent(
+        pushPayload: String,
+        completionHandler: @escaping (Result<[DeviceEvent], Error>) -> Void
+    )
+}
+
+public extension Notification.Name {
+    static let constellationStateUpdate = Notification.Name("constellationStateUpdate")
+}
+
+public struct ConstellationState {
+    public let localDevice: Device?
+    public let remoteDevices: [Device]
+}
+
+public class DeviceConstellation: DeviceConstellationProtocol {
+    var constellationState: ConstellationState?
+    let account: FirefoxAccount
+
+    init(account: FirefoxAccount) {
+        self.account = account
+    }
+
+    public func state() -> ConstellationState? {
+        return constellationState
+    }
+
+    public func refreshState() {
+        DispatchQueue.global().async {
+            FxALog.info("Refreshing device list...")
+            do {
+                let devices = try self.handleAccountAuthErrors { acct in
+                    try acct.fetchDevicesSync()
+                }
+                let localDevice = devices.first { $0.isCurrentDevice }
+                if localDevice?.subscriptionExpired ?? false {
+                    FxALog.debug("Current device needs push endpoint registration.")
+                }
+                let remoteDevices = devices.filter { !$0.isCurrentDevice }
+
+                let newState = ConstellationState(localDevice: localDevice, remoteDevices: remoteDevices)
+                self.constellationState = newState
+
+                FxALog.debug("Refreshed device list; saw \(devices.count) device(s).")
+
+                DispatchQueue.main.async {
+                    NotificationCenter.default.post(
```

Comment:
For my own education, why does this have to go out via notification center rather than e.g. returning whatever the equivalent of a `Promise` is here in swift land?

Reply:
I think most of the time the pattern used in these UIs such as the "send tab device list view" to set up listeners, launch `refreshState()` and forget about it. I'm very flexible on these choices: for example we removed device event observers that are present on Android Components.